### DISCUSSION TITLE
Embedded IDs Best Practice (MappedEntity vs ID)

### DISCUSSION THREAD

**Discussion by angus-clark** [micronaut-projects/micronaut-data#2957] [Q&A] [answered]
When creating an embedded id for a composite primary key should the fields that make up the embedded id be the raw values such as an Integer or UUID? Or should they be the full object/mapped entity in which they represent?

For example, say we have a database setup consisting of an "account" table and a "club" table and there is a many to many relationship exists between theses two tables. Therefore we have a third table "club_member" which connects accounts and clubs as represented below.

```
create table club_member
(
    club_id      uuid references club (id) on delete cascade not null,
    account_id   uuid references account (id) on delete cascade   not null,
    primary key (book_club_id, account_id)
);
```

So the question is what should the Embedded Id for this table look like out of the two options below?

Option 1:
```
@Serdeable
@Embeddable
public class ClubMemberId {

    @MappedProperty("account_id")
    private UUID accountId;

    @MappedProperty("club_id")
    private UUID clubId;
}
```

Option 2:
```
@Serdeable
@Embeddable
public class ClubMemberId {

    @MappedProperty("account_id")
    private Account account;

    @MappedProperty("club_id")
    private Club club;
}
```

**Comment by dstepanov** [comment, marked as answer]
Depends if you want in some cases JOIN an account or a club entity. See examples:

https://github.com/micronaut-projects/micronaut-data/blob/25a7a06398e7b895c34243d48a5287d31b3ff76a/data-tck/src/main/java/io/micronaut/data/tck/entities/ShipmentId.java#L26

https://github.com/micronaut-projects/micronaut-data/blob/25a7a06398e7b895c34243d48a5287d31b3ff76a/data-document-tck/src/main/java/io/micronaut/data/document/tck/entities/CountyPk.java#L23

**Reply by angus-clark** [reply]
Thank you! Its good to know they are both valid
